lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1538 @@
+/// `SubindexOp`, and `SubaccessOp`, and inserts multiplexers into connects to
+/// these subaccesses as well. Modifies the insertion location of the builder.
+/// Returns true if the `resetValue` was used in any way, false otherwise.
+static bool insertResetMux(ImplicitLocOpBuilder &builder, Value target,
+                           Value reset, Value resetValue) {
+  // Indicates whether the `resetValue` was assigned to in some way. We use this
+  // to erase unused subfield/subindex/subaccess ops on the reset value if they
+  // end up unused.
+  bool resetValueUsed = false;
+
+  for (auto &use : target.getUses()) {
+    Operation *useOp = use.getOwner();
+    builder.setInsertionPoint(useOp);
+    TypeSwitch<Operation *>(useOp)
+        // Insert a mux on the value connected to the target:
+        // connect(dst, src) -> connect(dst, mux(reset, resetValue, src))
+        .Case<ConnectOp, PartialConnectOp>([&](auto op) {
+          if (op.dest() != target)
+            return;
+          LLVM_DEBUG(llvm::dbgs() << "  - Insert mux into " << op << "\n");
+          auto muxOp = builder.create<MuxPrimOp>(reset, resetValue, op.src());
+          op.srcMutable().assign(muxOp);
+          resetValueUsed = true;
+        })
+        // Look through subfields.
+        .Case<SubfieldOp>([&](auto op) {
+          auto resetSubValue =
+              builder.create<SubfieldOp>(resetValue, op.fieldnameAttr());
+          if (insertResetMux(builder, op, reset, resetSubValue))
+            resetValueUsed = true;
+          else
+            resetSubValue.erase();
+        })
+        // Look through subindices.
+        .Case<SubindexOp>([&](auto op) {
+          auto resetSubValue =
+              builder.create<SubindexOp>(resetValue, op.indexAttr());
+          if (insertResetMux(builder, op, reset, resetSubValue))
+            resetValueUsed = true;
+          else
+            resetSubValue.erase();
+        })
+        // Look through subaccesses.
+        .Case<SubaccessOp>([&](auto op) {
+          auto resetSubValue =
+              builder.create<SubaccessOp>(resetValue, op.index());
+          if (insertResetMux(builder, op, reset, resetSubValue))
+            resetValueUsed = true;
+          else
+            resetSubValue.erase();
+        });
+  }
+  return resetValueUsed;
+}
+
+//===----------------------------------------------------------------------===//
+// Reset Network
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct ResetNode;
+
+enum class ResetKind { Uninferred, Async, Sync };
+
+/// A driver association between two IR values, given as a "source" value being
+/// driven onto a "destination" port or wire. Also contains location information
+/// for error reporting to the user.
+struct ResetDrive {
+  /// The output being driven.
+  ResetNode *dst;
+  /// The input node to the drive.
+  ResetNode *src;
+  /// The location to use for diagnostics.
+  Location loc;
+};
```

Comment:
This is cloning the IR into an on-the-side datastructure.  This probably makes sense in the Scala implementation, but MLIR provides use-def chains.  Can you just avoiding building this ResetDrive thing by using a pointer to an Operation?

Reply:
This is now reworked to use just track the IR nodes that make up the connection. The pass needs field granularity, so this is basically a (dstField, srcField) pair with an additional location pointing at the InstanceOp/ConnectOp that implies the conection.